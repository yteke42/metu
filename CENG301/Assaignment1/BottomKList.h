#ifndef BOTTOMKLIST_H
#define BOTTOMKLIST_H
template<typename T, int k>
class BottomKList {
private:
    T* arr[k];
public:
    BottomKList()
    {
        // MODIFY HERE: Implement the constructor.
        for(int i=0; i<k; i++){
            arr[i] = nullptr;
        }
        
    }
    ~BottomKList()
    {
        // MODIFY HERE: Implement the destructor (if needed).
        for(int i=0; i<k; i++) {
            delete arr[i];
        }

    }

    void addElement( T* element )
    {
        // MODIFY HERE: Implement the addElement function.
        //fiinding where to put the element

        int index = k-1;
        while (index >=0 && (arr[index] == nullptr || !(*arr[index] < *element))) {
            --index;
        }

        //shifting elements to right to insert the new one
        for (int j=k-1; j>index; --j) {
            arr[j] = arr[j-1];
        } 

        //finally, inserting
        arr[index+1] = element;

    }
    T* getBottomXthElement ( int x ) const  
    {
        // MODIFY HERE: Implement the getBottomXthElement function.
        // do not forget to return nullptr if x >= k or x < 0
        if(x <0 || x>=k) {
            return nullptr;
        }
        return arr[x];
    }
};
#endif