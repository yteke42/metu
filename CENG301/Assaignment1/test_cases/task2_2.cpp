#include <iostream>
using namespace std;

#include "BottomKList.h"

int main() {
    BottomKList<int,2> l1;
    int *a = new int(1),
        *b = new int(0),
        *c = new int(-2),
        *d = new int(-3),
        *e = new int(-7),
        *f = new int(-1);
    l1.addElement(a);
    // now l1.arr is [a, nullptr]
    l1.addElement(b);
    // now l1.arr is [b, a]

    int* l1_zero = l1.getBottomXthElement(0);
    int* l1_one = l1.getBottomXthElement(1);

    cout<<(*l1_zero)<<endl;
    cout<<(*l1_one)<<endl;

    l1.addElement(c);
    // now l1.arr is [c, b]

    l1_zero = l1.getBottomXthElement(0);
    l1_one = l1.getBottomXthElement(1);

    cout<<(*l1_zero)<<endl;
    cout<<(*l1_one)<<endl;

    l1.addElement(d);
    // now l1.arr is [d, c]

    l1_zero = l1.getBottomXthElement(0);
    l1_one = l1.getBottomXthElement(1);

    cout<<(*l1_zero)<<endl;
    cout<<(*l1_one)<<endl;

    l1.addElement(e);
    // now l1.arr is [e, d]

    l1_zero = l1.getBottomXthElement(0);
    l1_one = l1.getBottomXthElement(1);

    cout<<(*l1_zero)<<endl;
    cout<<(*l1_one)<<endl;

    l1.addElement(f);
    // now l1.arr is [e, d]

    l1_zero = l1.getBottomXthElement(0);
    l1_one = l1.getBottomXthElement(1);

    cout<<(*l1_zero)<<endl;
    cout<<(*l1_one)<<endl;

    BottomKList<int,2> l2;
    int *aa = new int(-1),
        *bb = new int(1);
    l2.addElement(aa);
    // now l2.arr is [a]
    l2.addElement(bb);
    // now l2.arr is [a]

    int* l2_zero = l2.getBottomXthElement(0);
    int* l2_one = l2.getBottomXthElement(1);

    cout<<(*l2_zero)<<endl;
    cout<<(*l2_one)<<endl;

    return 0;
}
