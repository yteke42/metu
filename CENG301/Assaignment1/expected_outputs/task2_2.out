0
1
-2
0
-3
-2
-7
-3
-7
-3
-1
1
